inlinino/instruments/hydroscat.py

```diff
@@ -0,0 +1,278 @@
+            ['%.9f' for n in range(2, len(cfg['variable_names']))]
+        cfg['terminator'] = b'\r\n'
+
+        self.output_cal_header = cfg["output_cal_header"]
+
+        # Active Timeseries Variables
+        self.active_variables = {var_name:True for var_name in cfg['variable_names']}
+        self.active_variables["Depth"] = False
+        self.active_variables["Voltage"] = False
+        self.widget_active_timeseries_variables_names = cfg['variable_names']
+        self.widget_active_timeseries_variables_selected = \
+                    [name for name in self.active_variables if self.active_variables[name]]
+                
+        # Update wavelengths for Spectrum Plot
+        # Plot is updated after the initial instrument setup or button click
+        # We only show values for bb channels since bb and fl channel numbers overlap
+        self.spectrum_plot_trace_names = ["beta"]
+        betawavs = [int(betalab[2:])
+                    for betalab in cfg['variable_names'] if betalab.startswith("bb")]
+        self.spectrum_plot_x_values = [np.array(sorted(betawavs))]
+
+        super().setup(cfg)
+
+        # If we were not previously in the idle state (i.e. we were
+        # RUNNING or STOPPED), transition to that state now
+        if self.previous_state != "IDLE":
+            try:
+                self.close()
+            except:
+                self.logger.warn("not previously idle")
+
+        # Set standard configuration and check cfg input
+        # super().setup(cfg)
+        self.logger.info("setup")
+
+    # State machine
+    #  start_state =>
+    #  IDLE =open button=> READY =start button=> START =/start command=>
+    #  RUNNING =stop button=> STOP =/stop command=> READY =close button=>
+    #  IDLE => ...
+
+    def change_state(self, state):
+        self.previous_state = self.state
+        self.state = state
+        self.logger.info("{} -> {}".format(self.previous_state, state))
+
+
+    def init_interface(self):
+        # invoked after channel open request (via Open button)
+        self.init()
+
+
+    def init(self):
+        self.hydroscat.date_command()
+        self.logger.info("DATE command")
+        self.hydroscat.flourescence_command()
+        self.logger.info("Fluorescence command")
+        self.hydroscat.burst_command()
+        self.logger.info("BURST command")
+        self.change_state("READY")
+
+
+    def write_to_interface(self):
+        if self.state == "START":
+            self.hydroscat.start_command()
+            if self.output_cal_header:
+                lines = self.hydroscat.header_lines()
+                for line in lines:
+                    self.logger.info(line)
+            self.logger.info("START command")
+            self.change_state("RUNNING")
+        elif self.state == "STOP":
+            self.hydroscat.stop_command()
+            self.logger.info("STOP command")
+            self.change_state("READY")
+
+
+    def log_start(self):
```

Comment:
Typically in Inlinino, instruments start sampling as soon as the connection with the instrument is open, when the user click on the 'Serial Port: Open' button. However, data is recorded only when the user start on 'Log Data: Start' the main reason is to make sure the instrument is measuring properly (e.g. no bubbles) before we record data to simplify data processing later on. 

For consistency with other instruments and if I understood correctly the code, I wonder if instead of overloading the `log_start` and `log_stop` methods, it wouldn't be more appropriate to overload the `open` and `close` methods.

Reply:
I like it! I puzzled over this one for awhile. I'm no longer overriding `log_start|stop()` or `open()` since hitting the `Open` button calls `init()` which now immediately sends a `START` command to HydroScat so that sampling begins straight away, as you suggest (with Start/Stop buttons just controlling logging which I had not fully appreciated). I also simplified the state machine in the process, removing the `READY` state. So, the only method I'm overriding now is `close()`.

Let me know what you think.